modules/support/src/bounded.rs

```diff
@@ -0,0 +1,183 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::Rate;
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::traits::Get;
+use primitives::{Balance, BlockNumber};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{CheckedSub, One, Zero},
+	FixedPointNumber, RuntimeDebug,
+};
+use sp_std::{marker::PhantomData, prelude::*, result::Result};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// The bounded type errors.
+#[derive(RuntimeDebug, PartialEq, Eq)]
+pub enum Error {
+	/// The value is out of bound.
+	OutOfBound,
+	/// The change diff exceeds the max absolute value.
+	ExceedMaxChangeAbs,
+}
+
+/// An abstract definition of bounded type. The type is within the range of `Range`
+/// and while update the inner value, the max absolute value of the diff is `MaxChangeAbs`.
+/// The `Default` value is minimum value of the range.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize), serde(transparent))]
+#[derive(Encode, PartialEq, Eq, PartialOrd, Ord, Copy, Clone, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[scale_info(skip_type_params(Range, MaxChangeAbs))]
+pub struct BoundedType<T: Encode + Decode, Range, MaxChangeAbs>(
+	T,
+	#[cfg_attr(feature = "std", serde(skip_serializing))] PhantomData<(Range, MaxChangeAbs)>,
+);
+
+impl<T: Encode + Decode + CheckedSub + PartialOrd, Range: Get<(T, T)>, MaxChangeAbs: Get<T>> Decode
+	for BoundedType<T, Range, MaxChangeAbs>
+{
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let inner = T::decode(input)?;
+		Self::try_from(inner).map_err(|_| "BoundedType: value out of bound".into())
+	}
+}
+
+impl<T: Default + Encode + Decode, Range: Get<(T, T)>, MaxChangeAbs: Get<T>> Default
+	for BoundedType<T, Range, MaxChangeAbs>
+{
+	fn default() -> Self {
+		let (min, _) = Range::get();
+		Self(min, PhantomData)
+	}
+}
+
+impl<T: Encode + Decode + CheckedSub + PartialOrd, Range: Get<(T, T)>, MaxChangeAbs: Get<T>>
+	BoundedType<T, Range, MaxChangeAbs>
+{
+	/// Try to create a new instance of `BoundedType`. Returns `Err` if out of bound.
+	pub fn try_from(value: T) -> Result<Self, Error> {
+		let (min, max) = Range::get();
+		if value < min || value > max {
+			return Err(Error::OutOfBound);
+		}
+		Ok(Self(value, PhantomData))
+	}
+
+	/// Set the inner value. Returns `Err` if out of bound or the diff with current value exceeds
+	/// the max absolute value.
+	pub fn try_set(&mut self, value: T) -> Result<(), Error> {
+		let (min, max) = Range::get();
+		let max_change_abs = MaxChangeAbs::get();
+		let old_value = &self.0;
+		if value < min || value > max {
+			return Err(Error::OutOfBound);
+		}
+
+		let abs = if value > *old_value {
+			value
+				.checked_sub(old_value)
+				.expect("greater number subtracting smaller one can't underflow; qed")
+		} else {
+			old_value
+				.checked_sub(&value)
+				.expect("greater number subtracting smaller one can't underflow; qed")
+		};
+		if abs > max_change_abs {
```

Comment:
we may want to directly set the value to min for some situations, but if the change exceeds the  max_change_abs when try_set min, whether to bypass this check or not? @xlc

Reply:
In this case you can make multiple calls in a batch, to gradually set the value until min is reached.